feat: Add slskd support to glueforward

digimotong opened the pull request:
## Description
See: https://github.com/slskd/slskd/issues/1011

This PR adds support for slskd to glueforward, allowing it to manage port forwarding for Soulseek clients through gluetun. The changes include:

- Added SlskdClient class for config file management
- Modified main application to support multiple service types
- Updated README with slskd documentation

## Configuration Notes
- Service types (qbittorrent/slskd) are mutually exclusive - only one can be configured per glueforward instance
- For multiple services, run separate containers with different SERVICE_TYPE values
- qBittorrent is the default service type, so existing configurations do not need to be edited

## Testing
- Tested with two parallel instances:
  - One configured for qBittorrent (existing functionality)
  - One configured for slskd (new functionality)
- Both instances working as intended with proper port forwarding
- Passed pylint with no issues

<details>
<summary>Click to view test screenshots</summary>

![glueforward-slskd logs](https://github.com/user-attachments/assets/e74c1c53-7d79-49fa-a928-4d2720b487c1)
![slskd Options Webpage](https://github.com/user-attachments/assets/16161963-0a18-442d-ba0a-7c1edf23fa41)

</details>

## Documentation
Updated README.md to include:
- slskd docker-compose example
- SERVICE_TYPE and SLSKD_CONFIG_PATH environment variables
- Note about required volume mount for slskd.yaml

## Other Notes
This is my first real PR, so I apologize if any mistakes were made! Please feel free to give feedback and let me know if anything needs changed/improved. Thanks! :)

GeoffreyCoulaud commented on glueforward/main.py:
```diff
@@ -35,6 +36,29 @@ def __optional_getenv(self, name: str) -> None | str:
 
+    def __init_service_client(self) -> None:
+        """Initialize the appropriate service client based on SERVICE_TYPE"""
+        service_type = getenv("SERVICE_TYPE", "qbittorrent").lower()
```
Great for retro compatibility, but we should still warn users if they're not setting `SERVICE_TYPE` in my opinion. I suggest using `__optional_getenv` and adding to it an optional default value (like getenv does)

digimotong replied:
Updated `__optional_getenv` to support default values and warn when not set. See https://github.com/GeoffreyCoulaud/glueforward/commit/5e177b931b980676a91094ceb75c089d21cbcb14

GeoffreyCoulaud commented on glueforward/slskd.py:
```diff
@@ -0,0 +1,58 @@
+            raise SlskdConfigError(f"Failed to update config: {str(e)}") from e
+
+
+class SlskdClient:
+    """Client for managing slskd port configuration"""
+
+    def __init__(self, config_path: str):
+        self.config = SlskdConfig(config_path)
+
+    def set_port(self, port: int) -> None:
+        """Update the listening port in slskd config"""
+        if 1024 <= port <= 65535:
+            self.config.set_port(port)
+        else:
+            logging.warning("Invalid port %d, defaulting to 50300", port)
+            self.config.set_port(50300)
```
Something that's implicit in here, is that both clients have a common interface (eg. `set_port(self, port: int) -> None`) and that could be explicited with an abstract base class.

```py
from abc import ABC, abstractmethod


class ServiceClient(ABC):

    @abstractmethod
    def set_port(self, port: int) -> None:
        """Set the exposed port for the service"""

# class SlskdClient(ServiceClient):
# ...
# class QBittorrentClient(ServiceClient):
# ...

```

digimotong replied:
Added an abstract base class for `ServiceClient`. See https://github.com/GeoffreyCoulaud/glueforward/commit/77fb04f12d44a5aa5e311533fce6f838224c0eec

GeoffreyCoulaud commented on glueforward/main.py:
```diff
@@ -75,18 +99,12 @@ def _setup(self) -> None:
-                "password": self.__required_getenv("QBITTORRENT_PASSWORD"),
-            },
-        )
+        self.__init_service_client()
```
I think it would be better if `__init_service_client` returned the client, and `_setup` was in charge of assigning it. It's more explicit in my opinion.  
Then, you could rename `__init_service_client` into `__create_service_client`.

digimotong replied:
Updated this when implementing the ServiceClient class (https://github.com/GeoffreyCoulaud/glueforward/commit/77fb04f12d44a5aa5e311533fce6f838224c0eec)

GeoffreyCoulaud commented on glueforward/slskd.py:
```diff
@@ -0,0 +1,58 @@
+    """Client for managing slskd port configuration"""
+
+    def __init__(self, config_path: str):
+        self.config = SlskdConfig(config_path)
```
Isn't there an option to make an HTTP call to slskd's webui?  
I haven't looked yet at their [swagger](https://github.com/slskd/slskd/blob/master/docs/config.md#features), but that may help

GeoffreyCoulaud commented on glueforward/slskd.py:
```diff
@@ -0,0 +1,58 @@
+
+    def set_port(self, port: int) -> None:
+        """Update the listening port in slskd config"""
+        if 1024 <= port <= 65535:
+            self.config.set_port(port)
+        else:
+            logging.warning("Invalid port %d, defaulting to 50300", port)
+            self.config.set_port(50300)
```
Why is there a range check? Is this a specificity of slskd?

digimotong replied:
Yes, I noticed this will actually cause slskd to error out and go into a boot loop. Here's the container logs when setting the listen_port to 0:

```
[12:26:33 WRN] Options (re)configuration rejected.

[12:26:33 WRN] Invalid configuration:

  Soulseek:

    The field ListenPort must be between 1024 and 65535.

[12:26:33 ERR] Error when dispatching 'OnConnectedAsync' on hub.

Microsoft.Extensions.Options.OptionsValidationException: A validation error has occurred.

   at Microsoft.Extensions.Options.OptionsFactory`1.Create(String name)

   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)

--- End of stack trace from previous location ---

   at System.Lazy`1.CreateValue()

   at Microsoft.Extensions.Options.OptionsCache`1.GetOrAdd[TArg](String name, Func`3 createOptions, TArg factoryArgument)

   at slskd.Core.API.ApplicationHub.OnConnectedAsync()

   at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.OnConnectedAsync(HubConnectionContext connection)

   at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.OnConnectedAsync(HubConnectionContext connection)

   at Microsoft.AspNetCore.SignalR.HubConnectionHandler`1.RunHubAsync(HubConnectionContext connection)
```

Totally open to changing how we handle this error. I just decided to default to what slskd has set up (50300).

GeoffreyCoulaud reviewed (changes requested):
Overall, good first contribution :) 

Just a few style changes to be made, and possibly a change to use slskd's webui instead of changing its config.

GeoffreyCoulaud commented on README.md:
```diff
@@ -96,13 +118,43 @@ services:
\ No newline at end of file
+- Service types are mutually exclusive (only one service per container instance). For multiple services, run separate containers with different SERVICE_TYPE values.
+- For slskd support:
+  - Swagger must be enabled
```
Does swagger need to be enabled? I thought it was needed only to see docs.

digimotong replied:
Forgot to double check this. Yeah, let me switch it off and test

digimotong replied:
Yep you're right! Swagger doesn't need to be on. I'll remove that reference

digimotong commented:
Super sorry, totally forgot I still had this marked as ready for review. You've probably been getting a ton of pings. I'll make sure to respond to these comments.

I also completely forgot to run a lint test, I'll make sure to double check that locally in a sec

And thank you for the feedback! It's much appreciated!

digimotong commented:
Ok, changes should now be ready for review (apologies for pushing earlier than intended).

Branch is now fully updated to use the slskd API instead of directly editing the config file. I've also added a shared ServiceClient class with methods for `set_port `and `_handle_request_exception`.

I understand this is a lot to look over, so no rush at all :) Let me know if anything should be updated. I appreciate the feedback!

GeoffreyCoulaud commented:
Awesome, and don't worry for the notifications, if I want to not be spammed I just tick off the subscription :)  
Don't hesitate to use the "request a review" button, that's the way I'm used to work. 

I'll take a look, but already it was a good idea to delegate some shared logic into `ServiceClient`.

GeoffreyCoulaud commented on README.md:
```diff
@@ -100,9 +147,12 @@ services:
 </table>
 
 1. Optional before gluetun v3.40.0, where all control server routes become private by default
+2. Required when SERVICE_TYPE=slskd
```
Suggestion : Add a similar clarification for `SERVICE_TYPE=qbittorrent`

digimotong replied:
Updated the README to add a similar disclaimer for qbittorrent.

I was kinda debating on if the qbittorrent variables should be marked as "Yes" or "No" in the Optional column. Let me know if that should be changed.

GeoffreyCoulaud commented on glueforward/slskd.py:
```diff
@@ -0,0 +1,140 @@
+    def __request(self, method: str, url: str, **kwargs) -> Optional[httpx.Response]:
+        """Send authenticated request to slskd API"""
+        headers = kwargs.pop('headers', {})
+        if self.__token:
```
I think we should use `if self.get_is_authenticated()` here

digimotong replied:
Agreed + updated.

GeoffreyCoulaud commented on glueforward/slskd.py:
```diff
@@ -0,0 +1,140 @@
+
+    def __update_config(self, config: dict) -> None:
+        """Update slskd config with new YAML"""
+        stream = StringIO()
+        yaml.dump(config, stream)
+        self.__request(
+            method="post",
+            url="/api/v0/options/yaml",
+            content=json.dumps(stream.getvalue()),
```
These conversion steps probably warrant a comment. I didn't know `ruamel.yaml`'s api, so I didn't understand at first why it needed a stream. 

Why do we need to json-dump it in the end, doesn't the buffer contain a string already?

digimotong replied:
This one was a little tricky! The API expects the raw YAML content wrapped in a JSON string, so json.dumps() ensures proper escaping of any special characters in the YAML. The StringIO stream lets us fully capture the YAML formatting/indentation. 

<details>
<summary>For reference, here's a screenshot of the swagger docs.</summary>

![api_v0_options_yaml](https://github.com/user-attachments/assets/e09e32cf-d853-4ae7-82bf-6214e99400fa)

</details>

I've also updated the code to add some comments for clarity (let me know if these should be adjusted).

GeoffreyCoulaud commented on glueforward/slskd.py:
```diff
@@ -0,0 +1,140 @@
+
+    def set_port(self, port: int) -> None:
+        """Update the listening port in slskd config via API"""
+        if not 1024 <= port <= 65535:
+            logging.warning("Invalid port %d, defaulting to 50300", port)
+            port = 50300
```
In my opinion, we should raise a retryable error instead, if what we wanted to do is not possible, then do nothing (no override with the default).

Something like `class SlskdIllegalPort(RetryableGlueforwardError)` would work.
Unfortunately, we cannot trigger a port reassignment on Gluetun, so we'll have to wait until a new port is assigned. I'm not really fond of that compromise, but that's the best we can do for now...

digimotong replied:
Agreed, it's probably better to leave the config as-is instead of updating it. Added the `SlskdIllegalPort` class for this.

GeoffreyCoulaud commented on glueforward/slskd.py:
```diff
@@ -0,0 +1,140 @@
+from errors import GlueforwardError, RetryableGlueforwardError
+from service_client import ServiceClient
+
+# Configure YAML parser
+yaml = YAML()
+yaml.preserve_quotes = True
+yaml.indent(mapping=2, sequence=4, offset=2)
```
The yaml parser is specific to the service, so we should make it a property setup during `__init__` to avoid polluting the exports.

digimotong replied:
Moved the yaml parser configuration into the ` __init__` method and updated all references.

GeoffreyCoulaud reviewed (changes requested):
Great work already, it just needs a final round of discussion and polishing.  
I didn't mention it before, but good idea using a comment-preserving yaml library!

Oh and, no need to mention the commits in the answers, I'll look at the changed files in whole, so I shouldn't miss it ;)

digimotong commented:
Thank you for the feedback! All the comments should now be addressed and ready for review. Please don't hesitate to let me know if anything needs updated/adjusted :)

GeoffreyCoulaud commented on glueforward/slskd.py:
```diff
@@ -0,0 +1,154 @@
+        # Configure YAML parser
+        self._yaml = YAML()
+        self._yaml.preserve_quotes = True
+        self._yaml.indent(mapping=2, sequence=4, offset=2)
```
Ah, my bad I should have specified, but I'd prefer if it were referenced above, after `__token` and renamed `__yaml`, to mark it [private-ish](https://en.wikipedia.org/wiki/Name_mangling#Python).

digimotong replied:
Good catch! Updated and renamed the YAML config.

GeoffreyCoulaud reviewed (commented):
Just a tiny little change, and we'll be good to go :)
Sorry for being picky about that :sweat_smile:

digimotong commented:
No problem at all! Happy to make any changes needed. Should now be ready for review again.

Feel free to let me know if anything else needs updated 👍

GeoffreyCoulaud commented:
Alright! Now it's merge time!
I'll let you squash your commits and choose the message + description. Then, that's going into `main` :smile:

digimotong committed:
Add slskd support to glueforward

- Implement SlskdClient class handling YAML configuration via API
- Add SERVICE_TYPE environment variable for service selection
- Introduce SLSKD_URL, SLSKD_USERNAME, SLSKD_PASSWORD variables
- Maintain backward compatibility with existing qBittorrent functionality
- Add input validation for port numbers (1024-65535)
- Document new functionality in README.md

The changes allow glueforward to manage port forwarding for slskd
users through VPN changes, matching existing qBittorrent capability.
All existing functionality remains unchanged.

digimotong commented:
Awesome!! Commits have been squashed. Ready for merge :)
